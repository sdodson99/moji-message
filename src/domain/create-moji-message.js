import { CharacterMatrixMap } from './character-matrix-map';

const CHARACTER_GAP = '\n\n\n';
const EMPTY_CELL_GAP = '     ';

/**
 * Convert a message to a Moji Message.
 * @param {string} message The message to convert.
 * @param {string} emoji The emoji to build the Moji Message from.
 * @returns The converted Moji Message.
 */
export function createMojiMessage(message, emoji) {
  const messageCharacters = [...message];

  const mojiMessageCharacters = messageCharacters.map((c) => createMojiCharacter(c, emoji));

  const mojiMessage = mojiMessageCharacters.join(CHARACTER_GAP);

  return mojiMessage;
}

/**
 * Convert a message character to a Moji Message character
 * @param {string} character A single character to convert.
 * @param {string} emoji The emoji to build the Moji Message character from.
 * @returns The converted Moji Message character.
 */
function createMojiCharacter(character, emoji) {
  const upperCaseCharacter = character.toUpperCase();

  const characterMatrix = CharacterMatrixMap[upperCaseCharacter];

  if (!characterMatrix) {
    return '';
  }

  const moijCharacterRows = characterMatrix.map((row) => createMojiCharacterRow(row, emoji));

  return moijCharacterRows.join('\n');
}

/**
 * Convert a character matrix row to the appropiate emoji representation.
 * @param {number[]} characterMatrixRow The row of character matrix cells.
 * @param {string} emoji The emoji to convert the character matrix to.
 * @returns The converted character matrix row value.
 */
function createMojiCharacterRow(characterMatrixRow, emoji) {
  const mojiCharacterCells = characterMatrixRow.map((cell) => createMojiCharacterCell(cell, emoji));

  return mojiCharacterCells.join('');
}

/**
 * Convert a character matrix cell to the appropiate emoji representation.
 * @param {number} characterMatrixCell The binary value in the character matrix cell.
 * @param {string} emoji The emoji to convert the character matrix to.
 * @returns The converted character matrix cell value.
 */
function createMojiCharacterCell(characterMatrixCell, emoji) {
  if (!characterMatrixCell) {
    return EMPTY_CELL_GAP;
  }

  return emoji;
}
