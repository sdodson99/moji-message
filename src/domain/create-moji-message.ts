import { CharacterMatrixMap } from './character-matrix-map';

export function createMojiMessage(
  message: string,
  messageEmoji: string,
  backgroundEmoji: string
) {
  const messageCharacters = [...message];

  const mojiMessageCharacters = messageCharacters.map((c) =>
    createMojiCharacter(c, messageEmoji, backgroundEmoji)
  );

  const mojiMessage = mojiMessageCharacters.join(
    createCharacterGap(backgroundEmoji)
  );

  return mojiMessage;
}

function createMojiCharacter(
  character: string,
  messageEmoji: string,
  backgroundEmoji: string
) {
  const upperCaseCharacter = character.toUpperCase();

  const characterMatrix = CharacterMatrixMap[upperCaseCharacter];

  if (!characterMatrix) {
    return '';
  }

  const moijCharacterRows = characterMatrix.map((row) =>
    createMojiCharacterRow(row, messageEmoji, backgroundEmoji)
  );

  return moijCharacterRows.join('\n');
}

function createMojiCharacterRow(
  characterMatrixRow: boolean[],
  messageEmoji: string,
  backgroundEmoji: string
) {
  const mojiCharacterCells = characterMatrixRow.map((cell) =>
    createMojiCharacterCell(cell, messageEmoji, backgroundEmoji)
  );

  return mojiCharacterCells.join('');
}

function createMojiCharacterCell(
  characterMatrixCell: boolean,
  messageEmoji: string,
  backgroundEmoji: string
) {
  if (!characterMatrixCell) {
    return backgroundEmoji;
  }

  return messageEmoji;
}

function createCharacterGap(emoji: string) {
  return `\n${emoji.repeat(5)}\n`;
}
