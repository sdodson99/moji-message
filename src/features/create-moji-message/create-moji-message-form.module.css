.formError {
  @apply text-red-700 text-sm;
}

.invalid {
  @apply border-red-700 outline-red-700;
}

.emojiPickerPopup {
  max-width: 80vw;
}
