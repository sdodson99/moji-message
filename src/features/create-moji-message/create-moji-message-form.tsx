import { createMojiMessage, CreateMojiMessageRequest } from '@/entities/moji-message';
import styles from './create-moji-message-form.module.css';
import classNames from 'classnames';
import { useForm } from 'react-hook-form';
import Picker from '@emoji-mart/react';
import Popup from 'reactjs-popup';
import { useState } from 'react';

type CreateMojiMessageFieldValues = {
  message: string;
  messageEmoji: string;
  backgroundEmoji: string;
};

type EmojiSelectEvent = {
  native: string;
};

export type CreateMojiMessageFormProps = {
  onMojiMessageCreate?: (mojiMessage: string, createMojiMessageRequest: CreateMojiMessageRequest) => void;
};

export function CreateMojiMessageForm({ onMojiMessageCreate }: CreateMojiMessageFormProps) {
  const {
    formState: { errors },
    register,
    watch,
    setValue,
    setFocus,
    handleSubmit,
  } = useForm<CreateMojiMessageFieldValues>({
    defaultValues: {
      backgroundEmoji: '⚪️',
    },
  });

  const messageField = register('message', {
    required: 'Please enter a message.',
  });

  const messageEmojiValue = watch('messageEmoji');
  const messageEmojiField = register('messageEmoji', {
    required: 'Please select an emoji.',
  });

  const backgroundEmojiValue = watch('backgroundEmoji');
  const backgroundEmojiField = register('backgroundEmoji', {
    required: 'Please select a background emoji.',
  });

  const [isMessageEmojiPickerOpen, setIsMessageEmojiPickerOpen] = useState(false);

  const onOpenMessageEmojiPickerClick = () => {
    setIsMessageEmojiPickerOpen(true);
  };

  const onMessageEmojiSelect = ({ native }: EmojiSelectEvent) => {
    setValue('messageEmoji', native, { shouldValidate: true });
    setIsMessageEmojiPickerOpen(false);
  };

  const onMessageEmojiPickerClose = () => {
    setIsMessageEmojiPickerOpen(false);
    setFocus('messageEmoji');
  };

  const [isBackgroundEmojiPickerOpen, setIsBackgroundEmojiPickerOpen] = useState(false);

  const onOpenBackgroundEmojiPickerClick = () => {
    setIsBackgroundEmojiPickerOpen(true);
  };

  const onBackgroundEmojiSelect = ({ native }: EmojiSelectEvent) => {
    setValue('backgroundEmoji', native, { shouldValidate: true });
    setIsBackgroundEmojiPickerOpen(false);
  };

  const onBackgroundEmojiPickerClose = () => {
    setIsBackgroundEmojiPickerOpen(false);
    setFocus('backgroundEmoji');
  };

  const onSubmit = (createMojiMessageFieldValues: CreateMojiMessageFieldValues) => {
    const mojiMessage = createMojiMessage(createMojiMessageFieldValues);

    onMojiMessageCreate?.(mojiMessage, createMojiMessageFieldValues);

    window.dataLayer?.push({
      event: 'convert_message',
      message_length: createMojiMessageFieldValues.message.length,
      emoji: createMojiMessageFieldValues.messageEmoji,
      background_emoji: createMojiMessageFieldValues.backgroundEmoji,
    });
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className="flex flex-col">
        <label htmlFor="message">
          Enter a <strong>Message</strong>
        </label>
        <input
          type="text"
          id="message"
          className={classNames('mt-1', {
            [styles.invalid]: errors.message,
          })}
          {...messageField}
        />
        {errors.message && (
          <p role="alert" className={classNames('mt-1', styles.formError)}>
            {errors.message.message}
          </p>
        )}
      </div>
      <div className="mt-4 flex flex-col">
        <label htmlFor="messageEmojiPicker">
          Choose an <strong>Emoji</strong>
        </label>
        <div>
          <button
            id="messageEmojiPicker"
            ref={messageEmojiField.ref}
            aria-label="Open emoji selector"
            className={classNames('mt-1 border rounded w-10 h-10 flex items-center justify-center', {
              [styles.invalid]: errors.messageEmoji,
            })}
            type="button"
            onClick={onOpenMessageEmojiPickerClick}
          >
            {messageEmojiValue ? (
              messageEmojiValue
            ) : (
              <svg
                xmlns="http://www.w3.org/2000/svg"
                xmlnsXlink="http://www.w3.org/1999/xlink"
                version="1.1"
                width="24"
                height="24"
                viewBox="0 0 24 24"
              >
                <path d="M12,17.5C14.33,17.5 16.3,16.04 17.11,14H6.89C7.69,16.04 9.67,17.5 12,17.5M8.5,11A1.5,1.5 0 0,0 10,9.5A1.5,1.5 0 0,0 8.5,8A1.5,1.5 0 0,0 7,9.5A1.5,1.5 0 0,0 8.5,11M15.5,11A1.5,1.5 0 0,0 17,9.5A1.5,1.5 0 0,0 15.5,8A1.5,1.5 0 0,0 14,9.5A1.5,1.5 0 0,0 15.5,11M12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20M12,2C6.47,2 2,6.5 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z" />
              </svg>
            )}
          </button>
          <Popup trigger={() => <span />} open={isMessageEmojiPickerOpen} onClose={onMessageEmojiPickerClose}>
            <Picker onEmojiSelect={onMessageEmojiSelect} autoFocus />
          </Popup>
        </div>
        {errors.messageEmoji && (
          <p role="alert" className={classNames('mt-1', styles.formError)}>
            {errors.messageEmoji.message}
          </p>
        )}
      </div>
      <div className="mt-4 flex flex-col">
        <label htmlFor="backgroundEmojiPicker">
          Choose a <strong>Background Emoji</strong>
        </label>
        <div>
          <button
            id="backgroundEmojiPicker"
            ref={backgroundEmojiField.ref}
            aria-label="Open background emoji selector"
            className={classNames('mt-1 border rounded w-10 h-10 flex items-center justify-center', {
              [styles.invalid]: errors.backgroundEmoji,
            })}
            type="button"
            onClick={onOpenBackgroundEmojiPickerClick}
          >
            {backgroundEmojiValue ? (
              backgroundEmojiValue
            ) : (
              <svg
                xmlns="http://www.w3.org/2000/svg"
                xmlnsXlink="http://www.w3.org/1999/xlink"
                version="1.1"
                width="24"
                height="24"
                viewBox="0 0 24 24"
              >
                <path d="M12,17.5C14.33,17.5 16.3,16.04 17.11,14H6.89C7.69,16.04 9.67,17.5 12,17.5M8.5,11A1.5,1.5 0 0,0 10,9.5A1.5,1.5 0 0,0 8.5,8A1.5,1.5 0 0,0 7,9.5A1.5,1.5 0 0,0 8.5,11M15.5,11A1.5,1.5 0 0,0 17,9.5A1.5,1.5 0 0,0 15.5,8A1.5,1.5 0 0,0 14,9.5A1.5,1.5 0 0,0 15.5,11M12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20M12,2C6.47,2 2,6.5 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2Z" />
              </svg>
            )}
          </button>
          <Popup trigger={() => <span />} open={isBackgroundEmojiPickerOpen} onClose={onBackgroundEmojiPickerClose}>
            <Picker onEmojiSelect={onBackgroundEmojiSelect} autoFocus />
          </Popup>
        </div>
        {errors.backgroundEmoji && (
          <p role="alert" className={classNames('mt-1', styles.formError)}>
            {errors.backgroundEmoji.message}
          </p>
        )}
      </div>
      <div className="mt-8">
        <button className="btn btn-primary">Convert Message</button>
      </div>
    </form>
  );
}
