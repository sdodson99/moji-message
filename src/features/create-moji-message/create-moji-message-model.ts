import {
  createMojiMessage as createMojiMessageBase,
  CreateMojiMessageRequest,
  CreateMojiMessageStore,
} from '../../domain';

export type CreateMojiMessageForm = {
  message?: string;
  messageEmoji?: string;
  backgroundEmoji?: string;
};

export class CreateMojiMessageModel {
  get createMessageForm() {
    return this._createMessageForm;
  }

  constructor(
    private createMojiMessageStore: CreateMojiMessageStore,
    private _createMessageForm: CreateMojiMessageForm = {}
  ) {}

  createMojiMessage() {
    const createMojiMessageRequest = {
      message: this.createMessageForm.message ?? '',
      messageEmoji: this.createMessageForm.messageEmoji ?? '',
      backgroundEmoji: this.createMessageForm.backgroundEmoji ?? '',
    };

    const mojiMessage = createMojiMessageBase(createMojiMessageRequest);
    this.logCreateMojiMessageEvent(createMojiMessageRequest);

    this.createMojiMessageStore.currentMojiMessageRequest =
      createMojiMessageRequest;

    return mojiMessage;
  }

  private logCreateMojiMessageEvent(
    createMessageRequest: CreateMojiMessageRequest
  ) {
    window.dataLayer?.push({
      event: 'convert_message',
      message_length: createMessageRequest.message.length,
      emoji: createMessageRequest.messageEmoji,
      background_emoji: createMessageRequest.backgroundEmoji,
    });
  }
}
