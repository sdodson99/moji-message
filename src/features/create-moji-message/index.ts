export { CreateMojiMessageForm } from './create-moji-message-form';
