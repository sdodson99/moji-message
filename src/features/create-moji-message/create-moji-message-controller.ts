import { CreateMojiMessageModel } from './create-moji-message-model';
import {
  CreateMojiMessageView,
  EmojiSelectEvent,
} from './create-moji-message-view';

export class CreateMojiMessageController {
  constructor(
    private createMojiMessageView: CreateMojiMessageView,
    private createMojiMessageModel: CreateMojiMessageModel
  ) {
    this.createMojiMessageView.addFormSubmitListener((e) => {
      this.handleMojiMessageFormSubmit(e);
    });

    this.createMojiMessageView.addMessageInputListener(() => {
      this.handleMessageInput();
    });

    this.createMojiMessageView.addMessageEmojiSelectListener((e) => {
      this.handleMessageEmojiSelect(e);
    });

    this.createMojiMessageView.addBackgroundEmojiSelectListener((e) => {
      this.handleBackgroundEmojiSelect(e);
    });
  }

  private handleMojiMessageFormSubmit(event: Event) {
    event.preventDefault();

    this.createMojiMessageView.setFormSubmitted();

    if (!this.validateMessageForm()) {
      return this.createMojiMessageView.focusFirstInvalidInput();
    }

    const mojiMessage = this.createMojiMessageModel.createMojiMessage();

    this.createMojiMessageView.setMojiMessageOutput(mojiMessage);
    this.createMojiMessageView.showMojiMessageOutputSection();
  }

  private handleMessageInput() {
    this.createMojiMessageModel.createMessageForm.message =
      this.createMojiMessageView.getMessageInputValue();

    this.validateMessage();
  }

  private handleMessageEmojiSelect({ emoji }: EmojiSelectEvent) {
    this.createMojiMessageView.showSelectedMessageEmoji(emoji);

    this.createMojiMessageModel.createMessageForm.messageEmoji = emoji;

    this.validateMessageEmoji();
  }

  private handleBackgroundEmojiSelect({ emoji }: EmojiSelectEvent) {
    this.createMojiMessageView.showSelectedBackgroundEmoji(emoji);
    this.createMojiMessageModel.createMessageForm.backgroundEmoji = emoji;
  }

  private validateMessageForm() {
    const messageValid = this.validateMessage();
    const emojiValid = this.validateMessageEmoji();

    return messageValid && emojiValid;
  }

  private validateMessage() {
    const messageInputValidity =
      this.createMojiMessageView.getMessageInputValidity();

    if (messageInputValidity.valueMissing) {
      this.createMojiMessageView.setMessageInputInvalid();
      this.createMojiMessageView.setMessageInputErrorMessage(
        'Please enter a message.'
      );
      return false;
    }

    if (!messageInputValidity.valid) {
      this.createMojiMessageView.setMessageInputInvalid();
      this.createMojiMessageView.setMessageInputErrorMessage(
        'Invalid message.'
      );
      return false;
    }

    this.createMojiMessageView.setMessageInputValid();
    this.createMojiMessageView.clearMessageInputErrorMessage();

    return true;
  }

  private validateMessageEmoji() {
    if (!this.createMojiMessageModel.createMessageForm.messageEmoji) {
      this.createMojiMessageView.setMessageEmojiPickerInvalid();
      this.createMojiMessageView.setMessageEmojiPickerErrorMessage(
        'Please select an emoji.'
      );
      return false;
    }

    this.createMojiMessageView.setMessageEmojiPickerValid();
    this.createMojiMessageView.clearMessageEmojiPickerErrorMessage();

    return true;
  }
}
