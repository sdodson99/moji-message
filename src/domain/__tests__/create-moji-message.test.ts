import { createMojiMessage } from '../create-moji-message';

describe('createMojiMessage', () => {
  let messageInput: string;
  let emoji: string;
  let backgroundEmoji: string;

  beforeEach(() => {
    messageInput = 'abc defghijklmnopqrstuvwxyz';
    emoji = '🔴';
    backgroundEmoji = '⚪️';
  });

  it('should return encoded message', () => {
    const mojiMessage = createMojiMessage(messageInput, emoji, backgroundEmoji);

    expect(mojiMessage).toMatchSnapshot();
  });

  it('should encode message with emoji', () => {
    const mojiMessage = createMojiMessage(messageInput, emoji, backgroundEmoji);

    expect(mojiMessage).toContain(emoji);
  });

  it('should encode message background with background emoji', () => {
    const mojiMessage = createMojiMessage(messageInput, emoji, backgroundEmoji);

    expect(mojiMessage).toContain(backgroundEmoji);
  });

  it('should return empty message for unknown characters', () => {
    const mojiMessage = createMojiMessage('%@#$', emoji, backgroundEmoji);

    expect(mojiMessage).not.toContain(emoji);
  });
});
