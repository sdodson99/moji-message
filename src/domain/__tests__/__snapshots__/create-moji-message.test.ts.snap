// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMojiMessage should return encoded message 1`] = `
"⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️🔴🔴🔴🔴⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️🔴🔴🔴⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️🔴🔴🔴⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️🔴⚪️⚪️
⚪️🔴⚪️🔴⚪️⚪️⚪️
⚪️🔴🔴⚪️⚪️⚪️⚪️
⚪️🔴⚪️🔴⚪️⚪️⚪️
⚪️🔴⚪️⚪️🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴⚪️🔴🔴⚪️
⚪️🔴⚪️🔴⚪️🔴⚪️
⚪️🔴⚪️🔴⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴⚪️⚪️🔴⚪️
⚪️🔴⚪️🔴⚪️🔴⚪️
⚪️🔴⚪️🔴⚪️🔴⚪️
⚪️🔴⚪️🔴⚪️🔴⚪️
⚪️🔴⚪️⚪️🔴🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️🔴🔴🔴⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️⚪️🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️🔴🔴⚪️
⚪️⚪️🔴🔴🔴🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴🔴🔴⚪️⚪️
⚪️🔴⚪️🔴⚪️⚪️⚪️
⚪️🔴⚪️⚪️🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️🔴⚪️🔴⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️🔴⚪️🔴⚪️
⚪️🔴⚪️🔴⚪️🔴⚪️
⚪️🔴🔴⚪️🔴🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️🔴⚪️🔴⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️🔴⚪️🔴⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️🔴⚪️⚪️⚪️🔴⚪️
⚪️⚪️🔴⚪️🔴⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️⚪️⚪️🔴⚪️
⚪️⚪️⚪️⚪️🔴⚪️⚪️
⚪️⚪️⚪️🔴⚪️⚪️⚪️
⚪️⚪️🔴⚪️⚪️⚪️⚪️
⚪️🔴⚪️⚪️⚪️⚪️⚪️
⚪️🔴🔴🔴🔴🔴⚪️
⚪️⚪️⚪️⚪️⚪️⚪️⚪️"
`;
